README.md
# Pixelant Front-End Starter Kit
### `generator-fe-kit`

[![npm version](https://badge.fury.io/js/generator-fe-kit.svg)](https://badge.fury.io/js/generator-fe-kit)
[![Build Status](https://travis-ci.org/pixelant/pixelant-fe-starter-kit.svg?branch=master)](https://travis-ci.org/pixelant/pixelant-fe-starter-kit)

![fe-kit](https://raw.githubusercontent.com/dmh/img/master/fe-kit.jpg)

### [CHANGELOG](https://github.com/pixelant/pixelant-fe-starter-kit/blob/master/CHANGELOG.md)
### [CONTRIBUTING](https://github.com/t3kit/t3kit/blob/master/CONTRIBUTING.md)


### Required dependencies:

- [Git](https://git-scm.com/)
- [NodeJs](http://nodejs.org/) >=6.0.0
- [NPM](https://github.com/npm/npm) >=3.9.0
- [Bower](http://bower.io/) >=1.7.7 `npm install -g bower`
- [Grunt-cli](http://gruntjs.com/) >=0.1.13 `npm install -g grunt-cli`
- [Yeoman](http://yeoman.io/) >=1.6.0 `npm install -g yo`

_Note: You might need to use `sudo` before `npm` command to install packages globally_
***

### Getting started with Pixelant Front-End Starter Kit

* For start you'll need to install Pixelant Front-End Starter Kit generator:

```sh
npm install -g generator-fe-kit
```

* Now that the generator is installed, create a directory for your new project:

```sh
mkdir name-of-project
cd name-of-project
```

* and then run:

```sh
yo fe-kit
```

* Next you will get questions and instructions in terminal. Need to choose type of project
  * Static site _(static sites, promo sites, one page applications)_
  * Front-End workflow _(experimental, AngularJS, React, Meteor, Mobile apps, etc...)_
  * felayout_t3kit _(Front-End layout for **t3kit** project)_
* Enter name of project
* Create repository on remote server
* Add repo link from created repository
* Initialize git repository
* Install dependencies

![fe-kit](https://raw.githubusercontent.com/dmh/img/master/fe-kit.gif)

### To update Pixelant Front-End Starter Kit generator run:

`npm update -g generator-fe-kit`

***

![Pixelant](https://raw.githubusercontent.com/dmh/img/master/P.png)

[Pixelant](http://www.pixelant.se/)


CHANGELOG.md

#### v1.9.0 `February 20, 2017`
- **[UPDATE]** felayout_bluemountain 1.1.0 ([fbf8866](https://github.com/pixelant/pixelant-fe-starter-kit/commit/fbf8866))
- **[UPDATE]** felayout_t3kit 1.8.0 ([8489690](https://github.com/pixelant/pixelant-fe-starter-kit/commit/8489690))

***

#### v1.8.0 `January 24, 2017`
- **[UPDATE]** felayout_t3kit 1.7.0 ([5c9b671](https://github.com/pixelant/pixelant-fe-starter-kit/commit/5c9b671))
- **[UPDATE]** felayout_bluemountain 1.0.0 ([359b490](https://github.com/pixelant/pixelant-fe-starter-kit/commit/359b490))

***

#### v1.7.0 `October 19, 2016`
- **[UPDATE]** felayout_bluemountain 0.6.0 ([8612269](https://github.com/pixelant/pixelant-fe-starter-kit/commit/8612269))
- **[UPDATE]** felayout_t3kit 1.6.0 ([e644b9e](https://github.com/pixelant/pixelant-fe-starter-kit/commit/e644b9e))

***

#### v1.6.0 `July 18, 2016`
- **[UPDATE]** felayout_t3kit 1.5.0 ([3d05b77](https://github.com/pixelant/pixelant-fe-starter-kit/commit/3d05b77))
- **[UPDATE]** felayout_bluemountain 0.5.0 ([5c908ec](https://github.com/pixelant/pixelant-fe-starter-kit/commit/5c908ec))
- **[TEST]** fix Travis CI tests ([7ba58d8](https://github.com/pixelant/pixelant-fe-starter-kit/commit/7ba58d8))
- **[TEST]** update tests, add bluemountain tests ([3c1fd4e](https://github.com/pixelant/pixelant-fe-starter-kit/commit/3c1fd4e))
- **[FEATURE]** add Github as an option, add felayout_bluemountain template ([fd1077a](https://github.com/pixelant/pixelant-fe-starter-kit/commit/fd1077a))
- **[CHORE]** update dependencies in README template ([618a86b](https://github.com/pixelant/pixelant-fe-starter-kit/commit/618a86b))
- **[FEATURE]** add felayout_bluemountain ([c730eb4](https://github.com/pixelant/pixelant-fe-starter-kit/commit/c730eb4))
- **[CHORE]** cleanup old configs and unneeded files ([3cfd911](https://github.com/pixelant/pixelant-fe-starter-kit/commit/3cfd911))
- **[UPDATE]** package.json dependencies ([a904a24](https://github.com/pixelant/pixelant-fe-starter-kit/commit/a904a24))
- **[CHORE]** disable force option in jscs and jshint ([f4e4ec4](https://github.com/pixelant/pixelant-fe-starter-kit/commit/f4e4ec4))
- **[DOC]** update README file ([cf7cd5f](https://github.com/pixelant/pixelant-fe-starter-kit/commit/cf7cd5f))
- **[DOC]** add CONTRIBUTING.md file ([abf3ee1](https://github.com/pixelant/pixelant-fe-starter-kit/commit/abf3ee1))
- **[FIX]** travis CI tool config ([ba88666](https://github.com/pixelant/pixelant-fe-starter-kit/commit/ba88666))
- **[UPDATE]** node.js (>=6.0.0) and npm (>=3.9.0) dependencies ([03b7956](https://github.com/pixelant/pixelant-fe-starter-kit/commit/03b7956))

***

#### v1.5.0 `July 8, 2016`
- **[UPDATE]** felayout_t3kit 1.4.0 ([cb68506](https://github.com/pixelant/pixelant-fe-starter-kit/commit/cb68506))
- **[DOC]** add license file (Podio feature_24) ([3cdae3e](https://github.com/pixelant/pixelant-fe-starter-kit/commit/3cdae3e))

***

#### v1.4.0 `June 23, 2016`
- **[UPDATE]** felayout_t3kit 1.3.0 ([5ec2943](https://github.com/pixelant/pixelant-fe-starter-kit/commit/5ec2943))

***
#### v1.3.0 `June 2, 2016`
- **[UPDATE]** felayout_t3kit 1.2.0 ([63cb460](https://github.com/pixelant/pixelant-fe-starter-kit/commit/63cb460))

#### v1.2.1 `May 13, 2016`
- **[UPDATE]** felayout_t3kit 1.1.1 ([3df0171](https://github.com/pixelant/pixelant-fe-starter-kit/commit/3df0171))

#### v1.2.0 `May 10, 2016`
- **[UPDATE]** felayout_t3kit 1.1.0 ([587c5cb](https://github.com/pixelant/pixelant-fe-starter-kit/commit/587c5cb))

#### v1.1.1 `2016-04-15`
- **[UPDATE]**  felayout_t3kit 1.1.0 ([649c278](https://github.com/pixelant/pixelant-fe-starter-kit/commit/649c278))

#### v1.1.0 `2016-03-23`
- **[UPDATE]**  felayout_t3kit 1.0.0 ([6972f53](https://github.com/pixelant/pixelant-fe-starter-kit/commit/6972f53))

#### v1.0.8 `2016-03-15`
- **[FIX]**  fix checking updates config ([f4353b4](https://github.com/pixelant/pixelant-fe-starter-kit/commit/f4353b4))

#### v1.0.7 `2016-03-15`
- **[UPDATE]**  update felayout_t3kit ([6b44ea7](https://github.com/pixelant/pixelant-fe-starter-kit/commit/6b44ea7))
- **[CHORE]**  add update-notifier config ([16cc4e3](https://github.com/pixelant/pixelant-fe-starter-kit/commit/16cc4e3))

#### v1.0.6 `2016-02-28`
- **[UPDATE]**  update felayout_t3kit ([c203a79](https://github.com/pixelant/pixelant-fe-starter-kit/commit/c203a79))
- **[DOC]**  update README file ([6960a66](https://github.com/pixelant/pixelant-fe-starter-kit/commit/6960a66))

#### v1.0.5 `2016-02-25`
- **[UPDATE]**  felayout_t3kit (fixing bugs) ([124006f](https://github.com/pixelant/pixelant-fe-starter-kit/commit/124006f))
- **[DOC]**  small updates to README file ([890ef4e](https://github.com/pixelant/pixelant-fe-starter-kit/commit/890ef4e))
- **[DOC]**  update README files ([b5ea0b6](https://github.com/pixelant/pixelant-fe-starter-kit/commit/b5ea0b6))

#### v1.0.4 `2016-02-17`
- **[FIX]**  extend ignored files list ([2cd9012](https://github.com/pixelant/pixelant-fe-starter-kit/commit/2cd9012))

#### v1.0.3 `2016-02-17`
- **[CHORE]**  update changelog script -> add new prefix UPDATE ([eb491c2](https://github.com/pixelant/pixelant-fe-starter-kit/commit/eb491c2))
- **[FIX]**  fix .gitignore bug in npm repo ([b050ed4](https://github.com/pixelant/pixelant-fe-starter-kit/commit/b050ed4))
- **[DOC]**  fix links in README file ([58bf5bf](https://github.com/pixelant/pixelant-fe-starter-kit/commit/58bf5bf))

#### v1.0.2 `2016-02-17`
- **[CHORE]**  changelog fixes ([7d109ca](https://github.com/pixelant/pixelant-fe-starter-kit/commit/7d109ca))
- **[FIX]**  fix install npm/bower dependencies prompt ([6359ff6](https://github.com/pixelant/pixelant-fe-starter-kit/commit/6359ff6))
- **[DOC]**  update documentation ([9c5ee50](https://github.com/pixelant/pixelant-fe-starter-kit/commit/9c5ee50))

#### v1.0.1 `2016-02-17`
- **[DOC]** update documentation ([5fc314e](https://github.com/pixelant/pixelant-fe-starter-kit/commit/5fc314e))

#### v1.0.0 `2016-02-16`

app/index.js
var generators = require('yeoman-generator');
var chalk = require('chalk');
var updateNotifier = require('update-notifier');
var ast = require('ast-query');
const boxen = require('boxen');

// Colors
var warning = chalk.bold.red;
var success = chalk.bold.green;
var yellow = chalk.bold.yellow;
// var blue = chalk.blue;
var gray = chalk.dim.gray;
var magenta = chalk.bold.magenta;
var inverse = chalk.inverse;

module.exports = generators.Base.extend({
    initializing: function() {
        if (process.argv.indexOf('--no-check-updates') === -1) {
            var done = this.async();
            const pkg = require('../package.json');
            const notifier = updateNotifier({
                pkg,
                updateCheckInterval: 0,
                callback: function(err, update) {
                    if (err) {
                        var networkError = boxen(warning('Network error!') + '\n' + 'Might be problems with internet connection or npm remote repo' + '\nRun ' + success('yo fe-kit --no-check-updates') + ' to skip update checking' + '\n' + 'Only if you are sure that you already have latest version.', {
                            padding: { top: 0, right: 2, bottom: 0, left: 2 },
                            margin:{ top: 0, right: 0, bottom: 1, left: 0 },
                            borderColor: 'red',
                            borderStyle: 'round'
                        });
                        console.log(networkError);
                        console.log(err);
                    } else {
                        var generatorNameMessage = boxen(gray('Pixelant Front-End Starter Kit (' + notifier.packageName + ')'), {
                            padding: { top: 0, right: 2, bottom: 0, left: 2 },
                            margin:{ top: 0, right: 0, bottom: 0, left: 0 },
                            borderColor: 'black',
                            borderStyle: 'round'
                        });
                        var noUpdatesMessage = boxen('You are using actual version of generator' + '\n' + success('Feel free to install and use any type of project'), {
                            padding: { top: 0, right: 2, bottom: 0, left: 2 },
                            margin:{ top: 0, right: 0, bottom: 1, left: 0 },
                            borderColor: 'green',
                            borderStyle: 'round'
                        });
                        if (update.latest !== update.current) {
                            var needToUpdateMessage = boxen(warning('Need to update before continuing!') + '\n' + 'Update available ' + gray(update.current) + ' → ' + magenta(update.latest) + ' \nRun ' + success('npm i -g ' + notifier.packageName) + ' to update', {
                                padding: { top: 0, right: 2, bottom: 0, left: 2 },
                                margin:{ top: 0, right: 0, bottom: 1, left: 0 },
                                borderColor: 'red',
                                borderStyle: 'round'
                            });
                            console.log(generatorNameMessage);
                            console.log(needToUpdateMessage);
                        } else {
                            console.log(generatorNameMessage);
                            console.log(noUpdatesMessage);
                        }
                        done();
                    }
                }
            });
        }
    },

    prompting: {
        projectType: function() {
            var done = this.async();
            var that = this;
            var prompts = [
                {
                    type: 'list',
                    name: 'projectType',
                    message: 'Project type:',
                    choices: [
                        {
                            name: gray('Static site'),
                            value: 'site'
                        },
                        {
                            name: gray('Front-End workflow (experimental)'),
                            value: 'fe'
                        },
                        {
                            name: 'felayout_t3kit',
                            value: 'felayout_t3kit'
                        },
                        {
                            name: 'felayout_bluemountain',
                            value: 'felayout_bluemountain'
                        }
                    ],
                },
                {
                    message: 'Is it right type of project?',
                    name: 'confirmed',
                    type: 'confirm',
                    default: true
                }
            ];
            var promptRecursive = function() {
                that.prompt(prompts, function(answers) {
                    if (answers.projectType === 'site' || answers.projectType === 'fe') {
                        that.log(answers.projectType + warning(' under construction. Choose another type'));
                        promptRecursive();
                    } else {
                        if (!answers.confirmed) {
                            promptRecursive();
                        } else {
                            that.projectType = answers.projectType;
                            done();
                        }
                    }
                });
            };
            promptRecursive();
        },

        prompts: function() {
            var done = this.async();
            var that = this;
            var prompts = [];
            //  ============================================
            //  General prompts ============================
            //  ============================================
            var projectName = {
                type: 'input',
                name: 'projectName',
                message: 'Project Name',
                default: this.appname
            };
            var git = {
                message: 'Do you want automatically initialize' + yellow(' git ') + 'repo\nand push it to the server?',
                name: 'autoGit',
                type: 'confirm',
                default: true,
                when: function(answer) {
                    return answer.sshLink !== 'skip' && answer.remoteType !== 'skip';
                }
            };
            var confirm = {
                message: 'Looks good?',
                name: 'confirmed',
                type: 'confirm',
                default: true
            };
            var repo = {
                type: 'input',
                name: 'repo',
                message: 'Need to create new repository on remote server',
                default: 'Done',
                validate: function(answer) {
                    if (answer === 'y' || answer === 'Y' || answer === 'yes' || answer === 'Yes' || answer === 'Done' || answer === 'done') {
                        return true;
                    }
                    return warning('You must create repo before continuing');
                },
                when: function(answer) {
                    return answer.remoteType !== 'skip';
                }
            };
            var remote = {
                type: 'list',
                name: 'remoteType',
                message: 'Remote repo server:',
                choices: [
                    {
                        name: 'Bitbucket',
                        value: 'bitbucket'
                    },
                    {
                        name: 'Github',
                        value: 'github'
                    },
                    {
                        name: 'Skip',
                        value: 'skip'
                    },
                ],
            };
            // var hook = {
            //     type: 'input',
            //     name: 'hook',
            //     message: 'Need to add ' + yellow('POST Hook') + ' to your project on Bitbucket\n' + blue('http://54.216.37.235/bitbucketpost.php'),
            //     default: 'Done',
            //     validate: function(answer) {
            //         if (answer === 'y' || answer === 'Y' || answer === 'yes' || answer === 'Yes' || answer === 'Done' || answer === 'done') {
            //             return true;
            //         }
            //         return warning('You must add POST Hook before continuing');
            //     }
            // };
            var bitbucketsshLink = {
                type: 'input',
                name: 'sshLink',
                message: 'SSH link of your new remote repository \n' + inverse('skip') + ' to skip this step',
                validate: function(answer) {
                    if (answer.length > 25 && answer.slice(0, 17) === 'git@bitbucket.org' && answer.slice(-4) === '.git' || answer === 'skip') {
                        return true;
                    }
                    return warning('Wrong repository link, try again...');
                },
                when: function(answer) {
                    return answer.remoteType === 'bitbucket';
                }
            };
            var githubsshLink = {
                type: 'input',
                name: 'sshLink',
                message: 'SSH link to your new remote repository\n' + inverse('skip') + ' to skip this step',
                validate: function(answer) {
                    if (answer.length > 22 && answer.slice(0, 14) === 'git@github.com' && answer.slice(-4) === '.git' || answer === 'skip') {
                        return true;
                    }
                    return warning('Wrong repository link, try again...');
                },
                when: function(answer) {
                    return answer.remoteType === 'github';
                }
            };

            var installDependencies = {
                message: 'Install npm/bower dependencies?',
                name: 'npmBower',
                type: 'confirm',
                default: true
            };

            //  ============================================
            //  felayout_t3kit =============================
            //  ============================================
            if (this.projectType === 'felayout_t3kit') {
                prompts.push(projectName);
                prompts.push(remote);
                prompts.push(repo);
                prompts.push(bitbucketsshLink);
                prompts.push(githubsshLink);
                prompts.push(git);
                prompts.push(installDependencies);
                prompts.push(confirm);
            }
            //  ********************************************
            //  --------------------------------------------

            //  ============================================
            //  felayout_bluemountain =============================
            //  ============================================
            if (this.projectType === 'felayout_bluemountain') {
                prompts.push(projectName);
                prompts.push(remote);
                prompts.push(repo);
                prompts.push(bitbucketsshLink);
                prompts.push(githubsshLink);
                prompts.push(git);
                prompts.push(installDependencies);
                prompts.push(confirm);
            }
            //  ********************************************
            //  --------------------------------------------

            //  ============================================
            //  prompt recursive  func =====================
            //  ============================================
            var promptRecursive = function() {
                that.prompt(prompts, function(answers) {
                    if (!answers.confirmed) {
                        promptRecursive();
                    } else {
                        var prop;
                        for (prop in answers) {
                            that[prop] = answers[prop];
                        }
                        done();
                    }
                });
            };
            promptRecursive();
        },
    },

    configuring: {
        parsePackageJson: function() {
            this.packageJson = this.fs.readJSON(this.templatePath() + '/' + this.projectType + '/package.json');
            this.packageJsonVersion = this.packageJson.version;
            this.packageJson.name = this.projectName;
            this.packageJson.version = '0.0.1';
            this.packageJson.description = 'Front-End layout for ' + this.projectName + ' project';
        },
        parseBowerJson: function() {
            this.bowerJson = this.fs.readJSON(this.templatePath() + '/' + this.projectType + '/bower.json');
            this.bowerJson.name = this.projectName;
        },
        parseGruntfile: function() {
            this.readGruntfile = this.fs.read(this.templatePath() + '/' + this.projectType + '/Gruntfile.js');
            this.gruntfileTree = ast(this.readGruntfile);
            if (this.remoteType === 'github') {
                this.gruntfileTree.var('remoteBranch').value('\'gh-pages\'');
            } else {
                this.gruntfileTree.var('remoteBranch').value('\'site\'');
            }
            this.gruntfileTree.var('remoteRepo').value('\'' + this.sshLink + '\'');
        },

    },

    // default: {
    // },

    writing: function() {
        this.fs.copy([
            this.templatePath() + '/' + this.projectType + '/**',
            this.templatePath() + '/' + this.projectType + '/**/.*',
            '!**/{Gruntfile.js, bower.json,package.json,.git,.npmignore,.gitignore,wct.conf.js,docs,test,README.md, LICENSE,CHANGELOG.md,CONTRIBUTING.md,.travis.yml}/**'],
            this.destinationPath()
        );
        this.fs.writeJSON(this.destinationPath('package.json'), this.packageJson);
        this.fs.writeJSON(this.destinationPath('bower.json'), this.bowerJson);
        this.fs.write(this.destinationPath('Gruntfile.js'), this.gruntfileTree.toString());
        this.fs.copyTpl(
            this.templatePath('README.md'),
            this.destinationPath('README.md'),
            { projectName: this.projectName,  projectType: this.projectType,  packageVersion: this.packageJsonVersion }
        );

        // Handle bug where npm has renamed .gitignore to .npmignore
        // https://github.com/npm/npm/issues/3763
        if (this.fs.exists(this.templatePath() + '/' + this.projectType + '/.npmignore')) {
            this.fs.copy(
                this.templatePath() + '/' + this.projectType + '/.npmignore',
                this.destinationPath('.gitignore')
            );
        } else {
            this.fs.copy(
                this.templatePath() + '/' + this.projectType + '/.gitignore',
                this.destinationPath('.gitignore')
            );
        }

        // if (this.projectType === 'felayout_t3kit') {
        //     this.fs.writeJSON( this.destinationPath('package.json'), this.packageJson);
        // }
    },
    install: {
        installDependencies: function() {
            if (this.npmBower) {
                this.npmInstall();
                this.bowerInstall();
            }
        },
        gitInit: function() {
            var that = this;
            if (this.autoGit && this.sshLink !== 'skip') {
                this.spawnCommand('git', ['init', '-q']).on('close', function() {
                    that.spawnCommand('git', ['add', '--all']).on('close', function() {
                        that.spawnCommand('git', ['commit', '-m', '"[INITIAL COMMIT]"', '-q']).on('close', function() {
                            that.spawnCommand('git', ['remote', 'add', 'origin', that.sshLink]).on('close', function() {
                                that.spawnCommand('git', ['push', '-u', '-q', 'origin', 'master']);
                            });
                        });
                    });
                });
            }
        },
    },
    end: {
        endLog: function() {
            if (!this.npmBower) {
                this.log(warning('\nYou need to install npm/bower dependencies manually\n') +
                    inverse('npm install\n' + 'bower install\n')
                );
            }
            if (!this.autoGit) {
                this.log(warning('\nYou need to initialize new git repo\nand push it to remote server manually\n') +
                    inverse('git init\n' +
                        'git add .\n' +
                        'git commit -m \'[INITIAL COMMIT]\'\n' +
                        'git remote add origin <<repo-link>>' + '\n' +
                        'git push -u origin master\n'
                        )
                    );
            }
            this.log(success('\nYou are all set now. ') + magenta('Happy coding!'));
        }
    }
});
